Forms: Rating field: Hearts variation + switch to SVGs

lezama opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Resolves FORMS-67

Follow-up to:
- https://github.com/Automattic/jetpack/pull/44094

<img width="1013" alt="Screenshot 2025-07-08 at 8 17 14 PM" src="https://github.com/user-attachments/assets/c3910534-4d0a-4c56-bf34-d5ac9611d531" />

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Adds a new “Hearts” style variation to the Rating field in Jetpack Forms.  
  * Introduces `hearts` glyph into `DEFAULT_GLYPHS` (♥).  
  * Registers the variation through the block’s `styles` array.  
  * Adds SCSS that swaps the default star glyph for a heart when the “Hearts” style is selected.  
* No behavioural changes to existing “Stars” style; it remains the default.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone. -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

1. In the block editor, add a **Form → Rating** field.
2. Select the nested **Rating input** block.
3. In the block toolbar or sidebar, open the **Styles** panel and choose **Hearts**.
4. Confirm the stars change to hearts in the editor.
5. Adjust “Highest rating” and “Default value” controls—confirm hearts render correctly.
6. Publish the post/page and view it on the front-end; hearts should display and behave the same as stars.

lezama committed:
the basics

lezama committed:
Refactor rating input

lezama committed:
changelog

lezama committed:
add styles back

lezama committed:
fix frontend styles

lezama committed:
fix frontend

lezama committed:
Update add-hearts-style-variation

ilonagl commented:
Left proposal for settings here: https://a8c.slack.com/archives/C052XEUUBL4/p1752582984507599?thread_ts=1752582780.967079&cid=C052XEUUBL4

lezama committed:
move settings to field block

lezama committed:
max value

lezama committed:
fix styles

lezama committed:
Refactor rating input to use SVG icons

Replaces Unicode star and heart characters with inline SVG icons for the rating input block, improving accessibility and styling flexibility. Updates JavaScript, SCSS, and PHP to support the new icon approach and removes unnecessary font imports and overrides.

lezama committed:
use a variation

lezama committed:
add to the toolbar

lezama committed:
better

lezama committed:
remove comments

simison commented on projects/packages/forms/src/contact-form/class-contact-form-field.php:
```diff
@@ -2088,10 +2089,29 @@ private function render_rating_field( $id, $label, $value, $class, $required, $r
+		}
+
+		$star_svg  = '<svg class="jetpack-field-rating__icon" viewBox="0 0 24 24" aria-hidden="true"><path d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.62L12 2 9.19 8.62 2 9.24l5.46 4.73L5.82 21z" fill="currentColor" stroke="var(--jetpack--contact-form--border-color, #8c8f94)" stroke-width="2" stroke-linejoin="round"></path></svg>';
+		$heart_svg = '<svg class="jetpack-field-rating__icon" viewBox="0 0 24 24" aria-hidden="true"><path d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z" fill="currentColor" stroke="var(--jetpack--contact-form--border-color, #8c8f94)" stroke-width="2" stroke-linejoin="roun ...
```
```
fill="currentColor" stroke="var(--jetpack--contact-form--border-color, #8c8f94)" stroke-width="2"
```
It might make sense to try move these inline styles to a CSS file, so that we don't make CSS specificty so high and harder to override.

simison commented on projects/packages/forms/src/blocks/input-rating/symbols.js:
```diff
@@ -1,26 +1,27 @@
 	return (
-		<div className="jetpack-field-rating__wrapper">
-			{ Array.from( Array( max ), ( _, i ) => i + 1 ).map( position => (
+		<div className="jetpack-field-rating__wrapper" aria-label="Rating control">
```
Translation function missing

simison commented on projects/packages/forms/src/blocks/input-rating/style.scss:
```diff
@@ -0,0 +1,45 @@
+		margin-bottom: 0;
+		cursor: pointer;
+		overflow: hidden;
+		font-family: "IBM Plex Sans JP", sans-serif;
```
Prolly not needed now with SVGs? edit: oh it was lavel :-) But is it needed anyway?

simison commented on projects/packages/forms/src/blocks/field-rating/edit.js:
```diff
@@ -32,9 +102,54 @@ export default function RatingFieldEdit( props ) {
+				<ToolbarGroup>
+					<ToolbarButton
+						icon={ lineSolid }
+						label={ __( 'Remove star', 'jetpack-forms' ) }
```
```suggestion
						label={
							variation === 'stars'
								? __( 'Remove star', 'jetpack-forms' )
								: __( 'Remove heart', 'jetpack-forms' )
						}
```

simison commented on projects/packages/forms/src/blocks/field-rating/edit.js:
```diff
@@ -32,9 +102,54 @@ export default function RatingFieldEdit( props ) {
+					/>
+					<ToolbarButton
+						icon={ plus }
+						label={ __( 'Add star', 'jetpack-forms' ) }
```
```suggestion
						label={
							variation === 'stars'
								? __( 'Add star', 'jetpack-forms' )
								: __( 'Add heart', 'jetpack-forms' )
						}
```

simison commented on projects/packages/forms/src/blocks/field-rating/index.js:
```diff
@@ -1,12 +1,19 @@
+} ) );
 const settings = {
 	...defaultSettings,
 	title: __( 'Rating (stars)', 'jetpack-forms' ),
```
```suggestion
	title: __( 'Rating field', 'jetpack-forms' ),
```

simison committed:
Update "star" rating to be more generic

simison commented on projects/packages/forms/src/blocks/input-rating/index.js:
```diff
@@ -1,18 +1,20 @@
 		default: { type: 'number', default: 0 },
+		variation: { type: 'string', default: 'stars' },
 	},
 	supports: {
```
I was first thinking it be good to add border colour support:

```js

		__experimentalBorder: {
			color: true,
			__experimentalDefaultControls: {
				color: true,
			},
		},
```


...but on quick test need to figure how to apply it to SVGs 😅 

<img width="1136" height="157" alt="Screenshot 2025-07-21 at 16 39 59" src="https://github.com/user-attachments/assets/fc5588b2-6c56-4979-bded-05539ddc1231" />

... on second thought, I think all we need is one colour, and either filled or unfilled stars. We don't really need the border at all? I think it complicates the symbol visually, and would make setting colour more complicated that it needs to be.

Typically all you want is something like this:

<img width="270" height="128" alt="Screenshot 2025-07-21 at 16 43 58" src="https://github.com/user-attachments/assets/9c334a1c-0293-43b1-8a27-6dbbfd1db02b" />

lezama replied:
went with one color :)

lezama committed:
add toolbar to both blocks

lezama committed:
cleanup styles

lezama committed:
works great

lezama committed:
cleanup

lezama committed:
reset

lezama committed:
fix style

lezama committed:
use Ilona's icons

lezama committed:
rating field

lezama committed:
Improve rating accessibility and toolbar labels

edanzer commented on projects/packages/forms/src/blocks/field-rating/index.js:
```diff
@@ -29,7 +57,13 @@ const settings = {
+			role: 'content',
+		},
 	},
+	styles: stylesArray,
```
I was just quick testing this. From what I can tell, the button the toolbar to toggle stars/hearts is working, but the styles button in the inspector controls on the sidebar is not. Clicking the 'Hearts' button there has no effect for me.

If I do update to hearts using the block toolbar, that works in the editor, and also renders correctly on the frontend. So all good there.

lezama replied:
fixed

edanzer commented on projects/packages/forms/src/blocks/shared/hooks/use-rating-sync.js:
```diff
@@ -0,0 +1,120 @@
+		relatedBlockId,
+		isParent,
+	};
+}
```
Seems like moving the settings from input to field, and syncing them, is a separate task from adding the stars vs hearts option? I wonder if these changes fcould be moved to a separate PR to keep this one simpler.

simison commented on projects/packages/forms/src/blocks/field-rating/edit.js:
```diff
@@ -28,13 +59,44 @@ export default function RatingFieldEdit( props ) {
 
-			{ /* Rating settings now moved to child rating-input block */ }
+			<InspectorControls>
+				<PanelBody title={ __( 'Rating settings', 'jetpack-forms' ) }>
```
```suggestion
				<PanelBody title={ __( 'Settings', 'jetpack-forms' ) }>
```

We can keep this simple, I think we call them just "Settings" elsewhere too.

lezama committed:
Refactor rating blocks to use context and className for style

Replaces the 'variation' attribute with className-based style selection and uses block context to share rating settings between field-rating and input-rating blocks. Removes unused hooks and utilities, updates toolbars to toggle styles via className, and simplifies attribute management for improved consistency and maintainability.

lezama committed:
cleanup

lezama committed:
Sync rating input className and simplify icon style check

In the input-rating block editor, the className from context is now synced to block attributes using useEffect. In the PHP field rendering, the logic for determining the heart style icon is simplified to check only field_classes, removing legacy and redundant checks.

lezama committed:
Sync rating default value changes between field and input

Introduces an onChangeDefault handler to the field-rating block and passes it to the input-rating block via context. Updates the Symbols component to toggle the rating value on click, ensuring that changes to the default value are reflected across both blocks.

lezama committed:
Fix SVG attribute casing in rating components

Replaced 'fill-rule' and 'clip-rule' with 'fillRule' and 'clipRule' in the SVG Path elements for both field-rating and input-rating blocks to ensure proper React compatibility and eliminate warnings.

lezama committed:
Refactor rating glyph property from 'char' to 'icon'

Renamed the 'char' property to 'icon' in DEFAULT_GLYPHS to better reflect its use as a React component, and updated references in the input-rating block accordingly. This improves code clarity and maintainability.

lezama committed:
Refactor rating toolbar style toggle logic

Introduces an isCurrentlyStars variable to simplify and clarify the logic for toggling between star and heart styles in the rating toolbar component.

dhasilva commented on projects/packages/forms/src/blocks/field-rating/style.scss:
```diff
@@ -1,53 +1,2 @@
-		}
-	}
-}
+@import '../input-rating/style';
```
This file is not used anywhere, right?

lezama replied:
It is! Somehow, it ensures that styles for the inner blocks get loaded. There may be a better way, but without this, the frontend breaks

lezama committed:
Add uniqueId to rating input for accessibility

Passes a uniqueId (clientId) from RatingInputEdit to Symbols to ensure unique aria-describedby, input id, and name attributes for each rating input instance. This improves accessibility and prevents conflicts when multiple rating inputs are present on the same page.

lezama committed:
fix editor styles

lezama committed:
improve styles

lezama committed:
Update projects/packages/forms/src/blocks/field-rating/edit.js

Co-authored-by: Mikael Korpela <mikael@ihminen.org>

lezama committed:
fix css files

lezama committed:
new lines

lezama committed:
new line

lezama committed:
this makes the build smaller?

ilonagl commented:
Great work, Miguel! 👏

A couple of notes to help optimize the first-time user experience:

Default size: Can we set the default size of the rating stars/hearts to XL? That would make them feel more in line and consistent with other input field sizes.

Default color: I know we currently default to the primary theme color, but most ratings—especially stars—are typically yellow. Would it make sense to offer that as the default instead? It aligns with common user expectations and avoids the need to manually search for a yellow shade. I realize this goes slightly against theme defaults, but it might be worth considering. Proposed yellow: #fdcc0d.

<img width="662" height="562" alt="Screenshot 2025-07-29 at 15 46 07" src="https://github.com/user-attachments/assets/c7e8da15-b054-4ebf-a366-d8f7b7e17f36" />
<img width="665" height="548" alt="Screenshot 2025-07-29 at 15 48 31" src="https://github.com/user-attachments/assets/0142cce7-4ee6-435c-a39a-42ad49e71b54" />

ilonagl commented:
<img width="653" height="134" alt="Screenshot 2025-07-29 at 15 32 31" src="https://github.com/user-attachments/assets/4b68ae1d-217f-4c8c-a086-3b71aff5f760" />

I noticed that on hover, the inside of the icon turns blue, but the outline feels slightly off. I get that there may be some limitations or quirks with how Gutenberg handles this, so it’s not a deal breaker—just flagging it in case it's something we can refine later.

ilonagl commented:
> Agreed, great idea! There might be some colour contrast checker function in Gutenberg to avoid colours which don't match with background, and rating colour could self-adjust by default? #fdcc0d basically works with dark/light backgrounds, but could cause issues with yellow/red/green etc bright backgrounds, so in those cases we either make the yellow brighter (close to white) or fallback to text colour?

We can fallback to text colour if that would be easier. It might be tricky to define the right yellow bright for all possible yellow/greeb/red backgrounds.
